### PR TITLE
Rely on file paths not assembly names

### LINKED ISSUE #83: Refactor RT to rely on assembly file path, not AssemblyName, for accuracy
There are cases where multiple referenced packages contain the same assembly (e.g. NewtonsoftJson.dll). Current implementation of RT causes all packages that contain the assembly to be marked as used, if the compiler reports this assembly as used. The idea is to refactor RT to rely on the actual file paths reported by the compiler and not translate them to assembly names.

### PR DESCRIPTION
**Pull request by stan-sz**
Fixed #83 for improved accuracy of detection.

Added test cases for projects with `ProduceReferenceAssembly` set

### CODE REVIEW

**Comment by stan-sz** [comment]
TODO: add a test case with reference assembly

**Review by erikmav — APPROVED** [review]

File: src/Analyzer/ReferenceTrimmerAnalyzer.cs [hunk 1]
```diff
@@ -87,7 +86,7 @@ private static void DumpUsedReferences(CompilationAnalysisContext context)
         {
             if (metadataReference.Display != null)
             {
-                string assemblyName = AssemblyName.GetAssemblyName(metadataReference.Display).Name;
+                string assemblyName = new Uri(metadataReference.Display).LocalPath;
```
Comment (dfederm) [inline src/Analyzer/ReferenceTrimmerAnalyzer.cs:89]: Why are you using a `Uri`?

**Comment by dfederm** [comment]
What about references like `<Reference Include="WindowsBase" />`?

File: src/Analyzer/ReferenceTrimmerAnalyzer.cs (hunk 1 above)
Comment (stan-sz) [reply src/Analyzer/ReferenceTrimmerAnalyzer.cs:89]: This is a path normalization function

File: src/Package/build/ReferenceTrimmer.targets [hunk 2]
```diff
@@ -18,13 +18,16 @@
       <_ReferenceTrimmerReferences Include="@(Reference)" />
       <_ReferenceTrimmerReferences Remove="@(_NETStandardLibraryNETFrameworkLib -> '%(FileName)')" />
 
+      <_ResolvedReferences Include="@(ReferencePath)" Condition="'%(ReferencePath.ReferenceSourceTarget)' == 'ResolveAssemblyReference'"/>
```
Comment (dfederm) [inline src/Package/build/ReferenceTrimmer.targets:21]: Nit: `_ReferenceTrimmerResolvedReferences`? To match the two surrounding this.

File: src/Analyzer/ReferenceTrimmerAnalyzer.cs (hunk 1 above)
Comment (dfederm) [reply src/Analyzer/ReferenceTrimmerAnalyzer.cs]: `Path.GetFullPath` should do the same thing?

File: src/Tasks/CollectDeclaredReferencesTask.cs [hunk 3]
```diff
@@ -97,31 +99,28 @@ public override bool Execute()
 
                     if (!string.IsNullOrEmpty(referenceHintPath) && File.Exists(referenceHintPath))
                     {
-                        referencePath = referenceHintPath;
+                        referencePath = Path.GetFullPath(referenceHintPath);
 
-                        // If a hint path is given and exists, use that assembly's name.
-                        referenceAssemblyName = AssemblyName.GetAssemblyName(referenceHintPath).Name;
+                        referenceAssemblyName = new Uri(referencePath).LocalPath;
```
Comment (dfederm) [inline src/Tasks/CollectDeclaredReferencesTask.cs:104]: `referenceAssemblyName` should be removed in favor of just `referencePath` now.

**Pull request merged**